tests/test_inference_problem.py

```diff
@@ -0,0 +1,253 @@
+
+
+"""
+###############################################################################
+
+                            USER CODE
+
+###############################################################################
+"""
+
+class MySensor(Sensor):
+    def __init__(self, name, position):
+        super().__init__(name, shape=(1,1))
+        self.position = position
+
+class MyForwardModel:
+    def __call__(self,parameter_list, sensors, time_steps):
+        """
+        evaluates 
+            fw(x, t) = A * x + B * t
+        """
+        A = parameter_list["A"]
+        B = parameter_list["B"]
+        
+        result = {}
+        for sensor in sensors:
+            result[sensor] = A * sensor.position + B * time_steps
+        return result
+
+    def parameter_list(self):
+        p = ParameterList()
+        p.define("A", None)
+        p.define("B", None)
+        return p
+
+
+class MyModelError(ModelError):
+    def __init__(self, fw, data):
+        self._fw = fw
+        self._ts, self._sensor_data = data
+
+    def __call__(self, prm):
+        sensors = self._sensor_data.keys()
+        model_response = self._fw(prm, sensors, self._ts)
+        error = {}
+        for sensor in sensors:
+            error[sensor] = model_response[sensor] - self._sensor_data[sensor]
+        return error
+
+class InferenceProblem:
+    def __init__(self):
+        self.latent = LatentParameterList()
+        self.noise = Noise()
+        self.model_errors = {}
+
+    def add_model_error(self, model_error, parameter_list, key=None):
+        key = key or len(self.model_errors)
+        assert key not in self.model_errors
+
+        self.model_errors[key] = model_error
+        self.latent.add_parameter_list(parameter_list, key)
+
+        return key
+
+"""
+###############################################################################
+
+                            MAIN CODE
+
+###############################################################################
+"""
+
+if __name__ == "__main__":
+    # Define the sensor
+    s1, s2, s3 = MySensor("S1", 0.2), MySensor("S2", 0.5), MySensor("S3", 42.)
+
+    fw = MyForwardModel()
+    prm = fw.parameter_list()
+    
+    # set the correct values
+    A_correct = 42.
+    B_correct = 6174.
+    prm["A"] = A_correct
+    prm["B"] = B_correct
+
+    np.random.seed= 6174
+
+    def generate_data(N_time_steps, noise_sd):
+        time_steps = np.linspace(0, 1, N_time_steps)
+        model_response = fw(prm, [s1, s2, s3], time_steps)
+        sensor_data = {}
+        for sensor, perfect_data in model_response.items():
+            sensor_data[sensor] = perfect_data + np.random.normal(0., noise_sd, N_time_steps) 
+        return time_steps, sensor_data
+
+    data1 = generate_data(101, noise_sd = 0.1)
+    data2 = generate_data(51, noise_sd = 0.2)
+
+    me1 = MyModelError(fw, data1)
+    me2 = MyModelError(fw, data2)
+
+    problem = InferenceProblem()
+    key1 = problem.add_model_error(me1, fw.parameter_list())
```

Comment:
I would suggest that the model error generates this list, not the forward model. This allows to retrieve in the model_error the parameter_list, but potentially modify values that are already given, e.g. the measured environmental temperature that we do not want to touch in the inference, or e.g. the transverse forces in Matthis compression tests that have been measured (but we don't want to infer). This model error could even by used with the same parameterized forward model to compute e.g. the tensile tests with/without transverse expansion. And here is another question - not directly related to the MR, but do we distinguish inputs to the forward model that are static (should/could be set when creating the model and are currently directly given in the constructor - not via a parameter_list but one by one) and those that are dynamic (stored in the parameter_list).

Reply:
First part: The parameter list generation/creation is not part of the interface. That means that there is no automatic treatment of the paramter lists, e.g. 
~~~py
for model_error in inference_problem._model_errors.values():
    prm = model_error.define_parameter_list()
    # now what?
~~~
That is because the user _has_ to touch this list to be forced to set values explicitly. Thus, it is his choice when/how to define it. So I agree, it can make sense for the user to implement his model_error such that it defines this list or implement it like I did before like:
~~~py
prm1 = fw.define_parameter_list() + model_error1.define_additional_parameters()
prm2 = fw.define_parameter_list() + model_error2.define_additional_parameters()
~~~

Long story short: It should not be forced when/how it is created.

Second part: I would not put too much effort into that because ultimately it is the users problem how to define his model/model_error.